File: main/analytics.js
const { dateKey } = require('./collector');

function sumDailyMap(daily, startDate, endDate) {
  let total = 0;
  for (const [day, count] of Object.entries(daily || {})) {
    if (day >= startDate && day <= endDate) total += count;
  }
  return total;
}

function findSnapshotOnOrBefore(snapshots, targetDate) {
  let best = null;
  for (const s of snapshots) {
    if (s.date <= targetDate) best = s;
    else break;
  }
  return best;
}

function daysAgo(n) {
  const d = new Date();
  d.setUTCDate(d.getUTCDate() - n);
  return d.toISOString().slice(0, 10);
}

function eachDay(startDate, endDate) {
  const out = [];
  const cur = new Date(`${startDate}T00:00:00.000Z`);
  const end = new Date(`${endDate}T00:00:00.000Z`);
  while (cur <= end) {
    out.push(cur.toISOString().slice(0, 10));
    cur.setUTCDate(cur.getUTCDate() + 1);
  }
  return out;
}

function deltaTotals(current, pastTotals) {
  if (!pastTotals) return { stars: null, downloads: null };
  return {
    stars: Math.max(0, (current.stars || 0) - (pastTotals.stars || 0)),
    downloads: Math.max(0, (current.downloads || 0) - (pastTotals.downloads || 0))
  };
}

function metricOrZero(value) {
  return Number.isFinite(value) ? value : 0;
}

/** Snapshot points for a totals field within [start, end]. */
function snapshotFieldPoints(snapshots, field, startDate, endDate) {
  return snapshots
    .filter((s) => s.date >= startDate && s.date <= endDate)
    .map((s) => ({
      date: s.date,
      value: s.totals?.[field] || 0
    }));
}

function repoMap(snapshot) {
  const map = new Map();
  for (const r of snapshot?.repos || []) {
    map.set(r.name, { stars: r.stars || 0, downloads: r.downloads || 0 });
  }
  return map;
}

function repoFieldDeltas(prevSnap, curSnap, field) {
  const prev = repoMap(prevSnap);
  const cur = repoMap(curSnap);
  const names = new Set([...prev.keys(), ...cur.keys()]);
  const changes = [];
  for (const name of names) {
    const a = prev.get(name)?.[field] || 0;
    const b = cur.get(name)?.[field] || 0;
    const delta = b - a;
    if (delta !== 0) changes.push({ name, delta });
  }
  return changes.sort((x, y) => Math.abs(y.delta) - Math.abs(x.delta));
}

/**
 * Daily deltas between consecutive snapshots in range, with per-repo breakdown.
 */
function snapshotDeltaPoints(snapshots, field, startDate, endDate) {
  const inRange = snapshots.filter((s) => s.date >= startDate && s.date <= endDate);
  if (!inRange.length) return [];

  const before = findSnapshotOnOrBefore(
    snapshots.filter((s) => s.date < startDate),
    startDate
  );

  const out = [];
  for (let i = 0; i < inRange.length; i++) {
    const prevSnap = i === 0 ? before : inRange[i - 1];
    const curSnap = inRange[i];
    const prevTotal = prevSnap?.totals?.[field]
      ?? (i === 0 && !before ? (curSnap.totals?.[field] || 0) : 0);
    const curTotal = curSnap.totals?.[field] || 0;
    const noBaseline = i === 0 && !before;
    const delta = noBaseline ? 0 : Math.max(0, curTotal - prevTotal);
    const repos = noBaseline || !prevSnap
      ? []
      : repoFieldDeltas(prevSnap, curSnap, field).filter((r) => r.delta > 0);
    out.push({ date: curSnap.date, value: delta, repos });
  }
  return out;
}

/**
 * Per-day star deltas from stargazer history, with which repos gained stars.
 */
function starHistoryDailyPoints(starHistories, startDate, endDate) {
  const byDay = new Map(); // day -> { value, repos: Map(name->count) }
  for (const [repo, hist] of Object.entries(starHistories || {})) {
    for (const [day, count] of Object.entries(hist.daily || {})) {
      if (day < startDate || day > endDate) continue;
      if (!byDay.has(day)) byDay.set(day, { value: 0, repos: new Map() });
      const entry = byDay.get(day);
      entry.value += count;
      entry.repos.set(repo, (entry.repos.get(repo) || 0) + count);
    }
  }
  return [...byDay.entries()]
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([date, entry]) => ({
      date,
      value: entry.value,
      repos: [...entry.repos.entries()]
        .map(([name, delta]) => ({ name, delta }))
        .sort((a, b) => b.delta - a.delta)
    }));
}

/**
 * Fill every day in range, carrying forward known cumulative values.
 * Before the first snapshot, uses that first known total so a brand-new
 * install still draws a full visible line across the selected window.
 */
function fillCarryForward(points, startDate, endDate) {
  if (!points.length) return [];
  const byDate = new Map(points.map((p) => [p.date, p.value]));
  let last = points[0].value;
  return eachDay(startDate, endDate).map((day) => {
    if (byDate.has(day)) last = byDate.get(day);
    return { date: day, value: last };
  });
}

/**
 * Fill every day in range for sparse daily deltas (0 on missing days),
 * starting at the first point date. Preserves `repos` metadata when present.
 */
function fillDailyZeros(points, startDate, endDate) {
  if (!points.length) return [];
  const byDate = new Map(points.map((p) => [p.date, p]));
  const firstDate = points[0].date;
  return eachDay(startDate, endDate)
    .filter((day) => day >= firstDate)
    .map((day) => {
      const hit = byDate.get(day);
      if (hit) return { date: day, value: hit.value || 0, repos: hit.repos || [] };
      return { date: day, value: 0, repos: [] };
    });
}

function mergeStarHistories(histories) {
  const merged = {};
  for (const hist of Object.values(histories || {})) {
    for (const [day, count] of Object.entries(hist.daily || {})) {
      merged[day] = (merged[day] || 0) + count;
    }
  }
  return merged;
}

function buildDashboard({
  snapshots,
  starHistories,
  currentTotals,
  rangeDays = 30
}) {
  const today = dateKey();
  const sorted = [...(snapshots || [])].sort((a, b) => a.date.localeCompare(b.date));

  let startDate;
  if (rangeDays === 'all') {
    startDate = sorted[0]?.date || daysAgo(90);
  } else {
    const n = rangeDays === 7 ? 7 : rangeDays === 90 ? 90 : 30;
    startDate = daysAgo(n);
  }

  const latestSnapshot = sorted.length ? sorted[sorted.length - 1] : null;
  const current = currentTotals?.stars != null
    ? { stars: currentTotals.stars, downloads: currentTotals.downloads }
    : (latestSnapshot?.totals || { stars: 0, downloads: 0 });
  const currentRepos = currentTotals?.repos || latestSnapshot?.repos || [];

  const mergedStars = mergeStarHistories(starHistories);
  const hasStarHistory = Object.keys(mergedStars).length > 0;

  // Baselines for period deltas: prefer a snapshot *before* the window.
  const beforeToday = findSnapshotOnOrBefore(
    sorted.filter((s) => s.date < today),
    daysAgo(1)
  );
  const before7d = findSnapshotOnOrBefore(sorted, daysAgo(7));
  const before30d = findSnapshotOnOrBefore(sorted, daysAgo(30));

  const starsTodayFromHistory = sumDailyMap(mergedStars, today, today);
  const stars7dFromHistory = sumDailyMap(mergedStars, daysAgo(7), today);
  const stars30dFromHistory = sumDailyMap(mergedStars, daysAgo(30), today);

  const dlToday = deltaTotals(current, beforeToday?.totals).downloads;
  const dl7d = deltaTotals(current, before7d?.totals).downloads;
  const dl30d = deltaTotals(current, before30d?.totals).downloads;
  const stToday = deltaTotals(current, beforeToday?.totals).stars;
  const st7d = deltaTotals(current, before7d?.totals).stars;
  const st30d = deltaTotals(current, before30d?.totals).stars;

  const metrics = {
    totalStars: current.stars || 0,
    totalDownloads: current.downloads || 0,
    starsToday: hasStarHistory ? starsTodayFromHistory : metricOrZero(stToday),
    stars7d: hasStarHistory ? stars7dFromHistory : metricOrZero(st7d),
    stars30d: hasStarHistory ? stars30dFromHistory : metricOrZero(st30d),
    downloadsToday: metricOrZero(dlToday),
    downloads7d: metricOrZero(dl7d),
    downloads30d: metricOrZero(dl30d)
  };

  // Star series: prefer stargazer history; fall back to snapshot totals/deltas.
  let starDailySeries;
  let starCumulativeSeries;
  if (hasStarHistory) {
    starDailySeries = fillDailyZeros(
      starHistoryDailyPoints(starHistories, startDate, today),
      startDate,
      today
    );
    const dayBeforeStart = (() => {
      const d = new Date(`${startDate}T00:00:00.000Z`);
      d.setUTCDate(d.getUTCDate() - 1);
      return d.toISOString().slice(0, 10);
    })();
    const beforeStart = sumDailyMap(mergedStars, '1970-01-01', dayBeforeStart);
    const daysInRange = Object.keys(mergedStars)
      .filter((d) => d >= startDate && d <= today)
      .sort();
    let running = beforeStart;
    const points = daysInRange.map((day) => {
      running += mergedStars[day];
      return { date: day, value: running };
    });
    if (!points.length && current.stars) {
      starCumulativeSeries = fillCarryForward(
        [{ date: today, value: current.stars }],
        startDate,
        today
      );
    } else {
      starCumulativeSeries = fillCarryForward(points, startDate, today);
    }
  } else {
    const starPoints = snapshotFieldPoints(sorted, 'stars', startDate, today);
    // If latest totals exist but no in-range snapshot point, seed with today.
    if (!starPoints.length && (current.stars || latestSnapshot)) {
      starPoints.push({
        date: latestSnapshot?.date || today,
        value: current.stars || 0
      });
    }
    starCumulativeSeries = fillCarryForward(starPoints, startDate, today);
    starDailySeries = fillDailyZeros(
      snapshotDeltaPoints(sorted, 'stars', startDate, today),
      startDate,
      today
    );
  }

  let downloadPoints = snapshotFieldPoints(sorted, 'downloads', startDate, today);
  if (!downloadPoints.length && (current.downloads || latestSnapshot)) {
    downloadPoints = [{
      date: latestSnapshot?.date || today,
      value: current.downloads || 0
    }];
  }
  const downloadTotalSeries = fillCarryForward(downloadPoints, startDate, today);
  const downloadDailySeries = fillDailyZeros(
    snapshotDeltaPoints(sorted, 'downloads', startDate, today),
    startDate,
    today
  );

  const repos = currentRepos.map((r) => {
    const hist = starHistories?.[r.name];
    const starDaily = hist?.daily || {};
    return {
      name: r.name,
      stars: r.stars,
      downloads: r.downloads,
      stars7d: hasStarHistory ? sumDailyMap(starDaily, daysAgo(7), today) : 0,
      stars30d: hasStarHistory ? sumDailyMap(starDaily, daysAgo(30), today) : 0
    };
  });

  return {
    metrics,
    series: {
      starsDaily: starDailySeries,
      starsCumulative: starCumulativeSeries,
      downloadsTotal: downloadTotalSeries,
      downloadsDaily: downloadDailySeries
    },
    repos,
    meta: {
      snapshotCount: sorted.length,
      hasStarHistory,
      rangeStart: startDate,
      rangeEnd: today,
      lastSnapshotAt: latestSnapshot?.timestamp || null
    }
  };
}

module.exports = {
  buildDashboard,
  mergeStarHistories,
  sumDailyMap,
  fillCarryForward,
  snapshotFieldPoints
};

File: crates/gh-stats-core/src/analytics.rs
use serde_json::{json, Map, Value};

use crate::collector::{date_key, HealthSnapshot};
use crate::store::Snapshot;
use crate::traffic::{aggregate_traffic_series, TrafficSnapshot};

fn days_ago(n: i64) -> String {
    (chrono::Utc::now() - chrono::Duration::days(n))
        .format("%Y-%m-%d")
        .to_string()
}

fn each_day(start: &str, end: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut cur = chrono::NaiveDate::parse_from_str(start, "%Y-%m-%d").ok();
    let end_d = chrono::NaiveDate::parse_from_str(end, "%Y-%m-%d").ok();
    while let (Some(c), Some(e)) = (cur, end_d) {
        if c > e {
            break;
        }
        out.push(c.format("%Y-%m-%d").to_string());
        cur = c.succ_opt();
    }
    out
}

fn sum_daily_map(daily: &Map<String, Value>, start: &str, end: &str) -> u64 {
    let mut total = 0u64;
    for (day, count) in daily {
        if day.as_str() >= start && day.as_str() <= end {
            total += count.as_u64().unwrap_or(0);
        }
    }
    total
}

fn find_snapshot_on_or_before<'a>(snapshots: &'a [Snapshot], target: &str) -> Option<&'a Snapshot> {
    let mut best = None;
    for s in snapshots {
        if s.date.as_str() <= target {
            best = Some(s);
        } else {
            break;
        }
    }
    best
}

fn fill_carry_forward(points: &[(String, u64)], start: &str, end: &str) -> Vec<Value> {
    if points.is_empty() {
        return vec![];
    }
    let by_date: std::collections::HashMap<_, _> =
        points.iter().map(|(d, v)| (d.clone(), *v)).collect();
    let mut last = points[0].1;
    each_day(start, end)
        .into_iter()
        .map(|day| {
            if let Some(v) = by_date.get(&day) {
                last = *v;
            }
            json!({ "date": day, "value": last })
        })
        .collect()
}

struct DailyPoint {
    date: String,
    value: u64,
    repos: Vec<Value>,
}

fn fill_daily_zeros(points: &[DailyPoint], start: &str, end: &str) -> Vec<Value> {
    if points.is_empty() {
        return vec![];
    }
    let by_date: std::collections::HashMap<_, _> = points
        .iter()
        .map(|p| (p.date.clone(), p))
        .collect();
    let first = points[0].date.clone();
    each_day(start, end)
        .into_iter()
        .filter(|d| d.as_str() >= first.as_str())
        .map(|day| {
            if let Some(hit) = by_date.get(&day) {
                json!({
                    "date": day,
                    "value": hit.value,
                    "repos": hit.repos.clone()
                })
            } else {
                json!({ "date": day, "value": 0, "repos": [] })
            }
        })
        .collect()
}

fn repo_field_deltas(prev: &Snapshot, cur: &Snapshot, field: &str) -> Vec<Value> {
    let mut prev_map = std::collections::HashMap::new();
    for r in &prev.repos {
        prev_map.insert(
            r.name.clone(),
            if field == "stars" { r.stars } else { r.downloads },
        );
    }
    let mut cur_map = std::collections::HashMap::new();
    for r in &cur.repos {
        cur_map.insert(
            r.name.clone(),
            if field == "stars" { r.stars } else { r.downloads },
        );
    }
    let mut names: std::collections::HashSet<_> = prev_map.keys().cloned().collect();
    names.extend(cur_map.keys().cloned());
    let mut changes: Vec<(String, i64)> = names
        .into_iter()
        .filter_map(|name| {
            let a = prev_map.get(&name).copied().unwrap_or(0) as i64;
            let b = cur_map.get(&name).copied().unwrap_or(0) as i64;
            let delta = b - a;
            if delta != 0 {
                Some((name, delta))
            } else {
                None
            }
        })
        .collect();
    changes.sort_by(|a, b| b.1.abs().cmp(&a.1.abs()));
    changes
        .into_iter()
        .map(|(name, delta)| json!({ "name": name, "delta": delta }))
        .collect()
}

fn merge_star_histories(histories: &Map<String, Value>) -> Map<String, Value> {
    let mut merged = Map::new();
    for hist in histories.values() {
        if let Some(daily) = hist.get("daily").and_then(|d| d.as_object()) {
            for (day, count) in daily {
                let n = merged.get(day).and_then(|v| v.as_u64()).unwrap_or(0)
                    + count.as_u64().unwrap_or(0);
                merged.insert(day.clone(), json!(n));
            }
        }
    }
    merged
}

fn snapshot_field_points(snapshots: &[Snapshot], field: &str, start: &str, end: &str) -> Vec<(String, u64)> {
    snapshots
        .iter()
        .filter(|s| s.date.as_str() >= start && s.date.as_str() <= end)
        .map(|s| {
            let value = match field {
                "stars" => s.totals.stars,
                _ => s.totals.downloads,
            };
            (s.date.clone(), value)
        })
        .collect()
}

fn snapshot_delta_points(snapshots: &[Snapshot], field: &str, start: &str, end: &str) -> Vec<DailyPoint> {
    let in_range: Vec<_> = snapshots
        .iter()
        .filter(|s| s.date.as_str() >= start && s.date.as_str() <= end)
        .collect();
    if in_range.is_empty() {
        return vec![];
    }
    let before_range: Vec<_> = snapshots
        .iter()
        .filter(|s| s.date.as_str() < start)
        .cloned()
        .collect();
    let before = find_snapshot_on_or_before(&before_range, start);
    let mut out = Vec::new();
    for (i, cur) in in_range.iter().enumerate() {
        let prev = if i == 0 {
            before
        } else {
            Some(in_range[i - 1])
        };
        let prev_total = prev
            .map(|p| if field == "stars" { p.totals.stars } else { p.totals.downloads })
            .unwrap_or(if i == 0 && before.is_none() {
                if field == "stars" {
                    cur.totals.stars
                } else {
                    cur.totals.downloads
                }
            } else {
                0
            });
        let cur_total = if field == "stars" {
            cur.totals.stars
        } else {
            cur.totals.downloads
        };
        let no_baseline = i == 0 && before.is_none();
        let delta = if no_baseline {
            0
        } else {
            cur_total.saturating_sub(prev_total)
        };
        let repos = if no_baseline {
            vec![]
        } else if let Some(p) = prev {
            repo_field_deltas(p, cur, field)
                .into_iter()
                .filter(|r| r.get("delta").and_then(|d| d.as_i64()).unwrap_or(0) > 0)
                .collect()
        } else {
            vec![]
        };
        out.push(DailyPoint {
            date: cur.date.clone(),
            value: delta,
            repos,
        });
    }
    out
}

fn star_history_daily_points(
    histories: &Map<String, Value>,
    start: &str,
    end: &str,
) -> Vec<DailyPoint> {
    let mut by_day: std::collections::BTreeMap<String, (u64, std::collections::HashMap<String, u64>)> =
        std::collections::BTreeMap::new();
    for (repo, hist) in histories {
        if let Some(daily) = hist.get("daily").and_then(|d| d.as_object()) {
            for (day, count) in daily {
                if day.as_str() >= start && day.as_str() <= end {
                    let n = count.as_u64().unwrap_or(0);
                    let entry = by_day.entry(day.clone()).or_insert_with(|| (0, std::collections::HashMap::new()));
                    entry.0 += n;
                    *entry.1.entry(repo.clone()).or_insert(0) += n;
                }
            }
        }
    }
    by_day
        .into_iter()
        .map(|(date, (value, repos_map))| {
            let mut repos: Vec<(String, u64)> = repos_map.into_iter().collect();
            repos.sort_by(|a, b| b.1.cmp(&a.1));
            DailyPoint {
                date,
                value,
                repos: repos
                    .into_iter()
                    .map(|(name, delta)| json!({ "name": name, "delta": delta }))
                    .collect(),
            }
        })
        .collect()
}

fn health_payload(health: Option<&HealthSnapshot>) -> Value {
    match health {
        Some(h) if h.fetched_at > 0 => json!({
            "fetchedAt": h.fetched_at,
            "issueCount": h.issues.len(),
            "issues": h.issues,
            "builds": h.builds,
            "releases": h.releases,
        }),
        _ => json!({
            "fetchedAt": null,
            "issueCount": 0,
            "issues": [],
            "builds": [],
            "releases": [],
        }),
    }
}

pub fn build_dashboard(
    snapshots: &[Snapshot],
    star_histories: &Map<String, Value>,
    range_days: Value,
    health: Option<&HealthSnapshot>,
    traffic: &[TrafficSnapshot],
) -> Value {
    let today = date_key();
    let mut sorted = snapshots.to_vec();
    sorted.sort_by(|a, b| a.date.cmp(&b.date));

    let start_date = if range_days.as_str() == Some("all") {
        sorted
            .first()
            .map(|s| s.date.clone())
            .unwrap_or_else(|| days_ago(90))
    } else {
        let n = range_days.as_u64().unwrap_or(30);
        let n = if n == 7 {
            7
        } else if n == 90 {
            90
        } else {
            30
        };
        days_ago(n)
    };

    let latest = sorted.last().cloned();
    let current_stars = latest.as_ref().map(|s| s.totals.stars).unwrap_or(0);
    let current_downloads = latest.as_ref().map(|s| s.totals.downloads).unwrap_or(0);
    let current_repos = latest.as_ref().map(|s| s.repos.clone()).unwrap_or_default();

    let merged = merge_star_histories(star_histories);
    let has_star_history = !merged.is_empty();

    let before_today_snaps: Vec<_> = sorted
        .iter()
        .filter(|s| s.date.as_str() < today.as_str())
        .cloned()
        .collect();
    let before_today = find_snapshot_on_or_before(&before_today_snaps, &days_ago(1));
    let before_7d = find_snapshot_on_or_before(&sorted, &days_ago(7));
    let before_30d = find_snapshot_on_or_before(&sorted, &days_ago(30));

    let metric = |cur: u64, past: Option<u64>| -> u64 {
        past.map(|p| cur.saturating_sub(p)).unwrap_or(0)
    };

    let total_npm: u64 = current_repos.iter().filter_map(|r| r.npm_downloads).sum();
    let total_pypi: u64 = current_repos.iter().filter_map(|r| r.pypi_downloads).sum();
    let total_crates: u64 = current_repos.iter().filter_map(|r| r.crate_downloads).sum();

    let traffic_agg = aggregate_traffic_series(traffic, &start_date, &today);
    let traffic_daily = traffic_agg.get("daily").cloned().unwrap_or(json!([]));
    let traffic_views_14d: u64 = traffic_daily
        .as_array()
        .map(|arr| {
            arr.iter()
                .filter_map(|d| d.get("views").and_then(|v| v.as_u64()))
                .sum()
        })
        .unwrap_or(0);
    let traffic_clones_14d: u64 = traffic_daily
        .as_array()
        .map(|arr| {
            arr.iter()
                .filter_map(|d| d.get("clones").and_then(|v| v.as_u64()))
                .sum()
        })
        .unwrap_or(0);

    let metrics = json!({
        "totalStars": current_stars,
        "totalDownloads": current_downloads,
        "starsToday": if has_star_history {
            sum_daily_map(&merged, &today, &today)
        } else {
            metric(current_stars, before_today.map(|s| s.totals.stars))
        },
        "stars7d": if has_star_history {
            sum_daily_map(&merged, &days_ago(7), &today)
        } else {
            metric(current_stars, before_7d.map(|s| s.totals.stars))
        },
        "stars30d": if has_star_history {
            sum_daily_map(&merged, &days_ago(30), &today)
        } else {
            metric(current_stars, before_30d.map(|s| s.totals.stars))
        },
        "downloadsToday": metric(current_downloads, before_today.map(|s| s.totals.downloads)),
        "downloads7d": metric(current_downloads, before_7d.map(|s| s.totals.downloads)),
        "downloads30d": metric(current_downloads, before_30d.map(|s| s.totals.downloads)),
        "npmDownloads": total_npm,
        "pypiDownloads": total_pypi,
        "crateDownloads": total_crates,
        "views14d": traffic_views_14d,
        "clones14d": traffic_clones_14d,
    });

    let (star_daily, star_cum) = if has_star_history {
        let daily = fill_daily_zeros(
            &star_history_daily_points(star_histories, &start_date, &today),
            &start_date,
            &today,
        );
        let mut running = 0u64;
        for (day, count) in &merged {
            if day.as_str() < start_date.as_str() {
                running += count.as_u64().unwrap_or(0);
            }
        }
        let mut points = Vec::new();
        let mut days: Vec<_> = merged.keys().filter(|d| d.as_str() >= start_date.as_str() && d.as_str() <= today.as_str()).cloned().collect();
        days.sort();
        for day in days {
            running += merged.get(&day).and_then(|v| v.as_u64()).unwrap_or(0);
            points.push((day, running));
        }
        if points.is_empty() && current_stars > 0 {
            (
                daily,
                fill_carry_forward(&[(today.clone(), current_stars)], &start_date, &today),
            )
        } else {
            (daily, fill_carry_forward(&points, &start_date, &today))
        }
    } else {
        let mut star_points = snapshot_field_points(&sorted, "stars", &start_date, &today);
        if star_points.is_empty() {
            if let Some(l) = &latest {
                star_points.push((l.date.clone(), current_stars));
            }
        }
        (
            fill_daily_zeros(
                &snapshot_delta_points(&sorted, "stars", &start_date, &today),
                &start_date,
                &today,
            ),
            fill_carry_forward(&star_points, &start_date, &today),
        )
    };

    let mut download_points = snapshot_field_points(&sorted, "downloads", &start_date, &today);
    if download_points.is_empty() {
        if let Some(l) = &latest {
            download_points.push((l.date.clone(), current_downloads));
        }
    }

    let repos: Vec<Value> = current_repos
        .iter()
        .map(|r| {
            let hist_daily = star_histories
                .get(&r.name)
                .and_then(|h| h.get("daily"))
                .and_then(|d| d.as_object())
                .cloned()
                .unwrap_or_default();
            json!({
                "name": r.name,
                "stars": r.stars,
                "downloads": r.downloads,
                "npmDownloads": r.npm_downloads,
                "pypiDownloads": r.pypi_downloads,
                "crateDownloads": r.crate_downloads,
                "stars7d": if has_star_history { sum_daily_map(&hist_daily, &days_ago(7), &today) } else { 0 },
                "stars30d": if has_star_history { sum_daily_map(&hist_daily, &days_ago(30), &today) } else { 0 },
            })
        })
        .collect();

    json!({
        "metrics": metrics,
        "series": {
            "starsDaily": star_daily,
            "starsCumulative": star_cum,
            "downloadsTotal": fill_carry_forward(&download_points, &start_date, &today),
            "downloadsDaily": fill_daily_zeros(
                &snapshot_delta_points(&sorted, "downloads", &start_date, &today),
                &start_date,
                &today
            ),
            "traffic": traffic_agg,
        },
        "repos": repos,
        "health": health_payload(health),
        "meta": {
            "snapshotCount": sorted.len(),
            "hasStarHistory": has_star_history,
            "hasTraffic": !traffic.is_empty(),
            "rangeStart": start_date,
            "rangeEnd": today,
            "lastSnapshotAt": latest.map(|s| s.timestamp)
        }
    })
}
